### PR TITLE
disable spherical harmonics when using cubemap textures

### PR DESCRIPTION
**Pull request by akaszynski**
VTK issues the following warning for any examples using cubemap textures:

```
vtkOpenGLRenderer.cxx:250   WARN| vtkOpenGLRenderer (0x32be840): Cannot compute spherical harmonics of a cubemap, fall back to irradiance texture
```

For example, when running:

```py
from pyvista import examples
import pyvista as pv
pl = pv.Plotter(lighting=None)
dataset = examples.download_sky_box_cube_map()
pl.add_mesh(pv.Sphere(), pbr=True, metallic=0.9, roughness=0.2)
pl.set_environment_texture(dataset)
pl.camera_position = 'xy'
pl.show()
```

The solution, is provided in the comment section of [Introducing Physically Based Rendering with VTK](https://www.kitware.com/vtk-pbr/):

> If one is using a cubemap texture (and not a traditional texture), one should set UseSphericalHarmonicsOff on the texture as well.

---

This fix requires changing the base class of `pyvista.Renderer` from `vtk.vtkRenderer` to `vtk.vtkOpenGLRenderer` as `UseSphericalHarmonicsOff` is available only in `vtk.vtkOpenGLRenderer`.

Also, the example was lacking, so it's been improved as well.

### CODE REVIEW

**Review by adeak — COMMENTED** [review]
> This fix requires changing the base class of `pyvista.Renderer` from `vtk.vtkRenderer` to `vtk.vtkOpenGLRenderer`

Firstly, I think this should be safe, because `vtkOpenGLRenderer` is a direct subclass of `vtkRenderer`:
```py
>>> vtk.vtkOpenGLRenderer.mro()[1:] == vtk.vtkRenderer.mro()
True
```
(the first item in the MRO is `vtk.vtkOpenGLRenderer` itself)

Secondly, this is interesting: [the docs says](https://vtk.org/doc/nightly/html/classvtkOpenGLRenderer.html#details)

> vtkOpenGLRenderer is a concrete implementation of the abstract class vtkRenderer. vtkOpenGLRenderer interfaces to the OpenGL graphics library.

So apparently `vtkRenderer` is an abstract class, and sure enough on my system I see this:
```py
>>> vtk.vtkRenderer()
<vtkmodules.vtkRenderingOpenGL2.vtkOpenGLRenderer(0x186ddd0) at 0x7f05ac338460>
>>> vtk.vtkOpenGLRenderer()
<vtkmodules.vtkRenderingOpenGL2.vtkOpenGLRenderer(0x1868dd0) at 0x7f05ac3384c0>
```

The only thing I'm wondering is whether anyone downstream might be using non-OpenGL renderers... is this possible?

File: pyvista/plotting/renderer.py [hunk 1]
```diff
@@ -2613,16 +2613,41 @@ def set_environment_texture(self, texture):
...
         Parameters
         ----------
         texture : vtk.vtkTexture
             Texture.
+
+        Warnings
+        --------
+        This texture must be expressed in linear color space. If the
+        texture is in sRGB color space, set the color flag on the texture or
+        set the argument isSRGB to true.
```
[the remark this answers is not in the document]
Comment (adeak) [reply pyvista/plotting/renderer.py:2627]: I see. [The VTK method indeed has that kwarg](https://vtk.org/doc/nightly/html/classvtkOpenGLRenderer.html#a2757170684619e8463a3ea6c6259d85e). We should either expose it or remove the doc reference.

[the remark this answers is not in the document]
Comment (akaszynski) [reply pyvista/plotting/renderer.py]: Done in 0f34c21.

**Comment by akaszynski** [comment]
> The only thing I'm wondering is whether anyone downstream might be using non-OpenGL renderers... is this possible?

I guess this is possible, but I'm not seeing this to be the case for xvfb, EGL (OSMesa) and we test on a lot of platforms as is. Let's put it out and if there are any issues we can always revert with a fast patch release.

File: pyvista/plotting/renderer.py [hunk 2]
```diff
@@ -2606,25 +2606,47 @@ def set_background(self, color, top=None):
...
-        must be expressed in linear color space. If the texture is in
-        sRGB color space, set the color flag on the texture or set the
-        argument isSRGB to true.
+        constructed with a valid ``vtk.vtkImageData``.
 
         Parameters
         ----------
         texture : vtk.vtkTexture
             Texture.
+        is_srgb : bool, optional
```
Comment (adeak) [inline pyvista/plotting/renderer.py:2622]:
```suggestion

        is_srgb : bool, optional
```

File: pyvista/plotting/renderer.py [hunk 3]
```diff
@@ -2606,25 +2606,47 @@ def set_background(self, color, top=None):
...
+            If the texture is in sRGB color space, set the color flag on the
+            texture or set the parameter this parameter to ``True``.
```
Comment (adeak) [inline pyvista/plotting/renderer.py:2624]:
```suggestion
            texture or set this parameter to ``True``. Textures are assumed
            to be in linear color space by default.
```

File: tests/plotting/test_plotting.py [hunk 4]
```diff
@@ -244,6 +245,20 @@ def test_pbr(sphere):
     pl.show(before_close_callback=verify_cache_image)
 
 
+@skip_not_vtk9
+@skip_windows
+@skip_mac
+@pytest.mark.skipif(CI_WINDOWS, reason="Windows CI testing segfaults on pbr")
```
Comment (adeak) [inline tests/plotting/test_plotting.py:251]: Does it make sense to have both `@skip_windows` and a manual skipif on `CI_WINDOWS`? I could imagine that CI_WINDOWS would be more informative due to the segfault message, but shouldn't that decorator come first then? Not sure.

File: tests/plotting/test_plotting.py [hunk 5]
```diff
@@ -71,6 +71,7 @@
 # TODO: once we have a stable release for VTK, remove these.
 HIGH_VARIANCE_TESTS = {
     'test_pbr',
+    'test_set_environment_texture',
```
Comment (adeak) [inline tests/plotting/test_plotting.py:74]: Is this missing the last word?
```suggestion
    'test_set_environment_texture_cubemap',
```

File: tests/plotting/test_plotting.py (hunk 4 above)
Comment (akaszynski) [reply tests/plotting/test_plotting.py:251]:
```suggestion
@skip_windows
@skip_mac
```

We really only need one skip windows.

File: tests/plotting/test_plotting.py [hunk 6]
```diff
@@ -244,6 +245,19 @@ def test_pbr(sphere):
     pl.show(before_close_callback=verify_cache_image)
 
 
+@skip_not_vtk9
```
Comment (adeak) [inline tests/plotting/test_plotting.py:248]:
```suggestion
@skip_9_1_0
```

**Review by adeak — APPROVED** [review]
Approving, hopefully CI will pass now that we're skipping the new test for 9.0.3.

File: pyvista/plotting/renderer.py [hunk 7]
```diff
@@ -2606,25 +2606,49 @@ def set_background(self, color, top=None):
...
+        >>> _ = pl.add_mesh(pv.Sphere(), pbr=True, metallic=0.9, roughness=0.4)
+        >>> pl.set_environment_texture(cubemap)
+        >>> pl.camera_position = 'xy'
+        >>> pl.show()
+
         """
+        # cube_map textures cannot use sphrical harmonics
+        if texture.cube_map:
+            self.AutomaticLightCreationOff()
+            self.UseSphericalHarmonicsOff()
```
Comment (akaszynski) [inline pyvista/plotting/renderer.py:2648]:
```suggestion
            # disable spherical harmonics was added in 9.1.0
            if hasattr(self, 'UseSphericalHarmonicsOff'):
                self.UseSphericalHarmonicsOff()
```

File: pyvista/plotting/renderer.py [hunk 8]
```diff
@@ -2606,25 +2606,49 @@ def set_background(self, color, top=None):
...
+        >>> import pyvista as pv
+        >>> pl = pv.Plotter(lighting=None)
+        >>> cubemap = examples.download_sky_box_cube_map()
+        >>> _ = pl.add_mesh(pv.Sphere(), pbr=True, metallic=0.9, roughness=0.4)
+        >>> pl.set_environment_texture(cubemap)
+        >>> pl.camera_position = 'xy'
+        >>> pl.show()
+
         """
+        # cube_map textures cannot use sphrical harmonics
```
Comment (akaszynski) [inline pyvista/plotting/renderer.py:2645]:
```suggestion
        # cube_map textures cannot use spherical harmonics
```

File: pyvista/plotting/renderer.py (hunk 7 above)
Comment (adeak) [reply pyvista/plotting/renderer.py]: Derp, we can then add back the new test for 9.0.3. Sorry.

File: tests/plotting/test_plotting.py [hunk 9]
```diff
@@ -244,6 +245,19 @@ def test_pbr(sphere):
     pl.show(before_close_callback=verify_cache_image)
 
 
+@skip_9_1_0
```
Comment (adeak) [inline tests/plotting/test_plotting.py:248]:
```suggestion
@kip_not_vtk9
```

File: tests/plotting/test_plotting.py [hunk 10]
```diff
@@ -244,6 +245,19 @@ def test_pbr(sphere):
     pl.show(before_close_callback=verify_cache_image)
 
 
+@kip_not_vtk9
```
Comment (adeak) [inline tests/plotting/test_plotting.py:248]: Time for bed, sorry.
```suggestion
@skip_not_vtk9
```

**Pull request merged**